### NLP Task

4 data sets of Sentimental Analysis

tweets
* 1)litigious
*2)negative
*3)positive
*4)uncertainty

To Do Steps:
   * 1) Try to convert your csv files into dataframe with both  features and labels
   * 2) Apply text preprocessing on the given data set
   * 3) Apply Bag of words and TF-IDF

In [45]:
import pandas as pd
import numpy as np

In [3]:
litigious=pd.read_csv(r"D:\Data Sets\ML Assignment NLP\dataset\litigious.csv",encoding="latin-1")
litigious

Unnamed: 0                                               Text Language
0                0  @Charlie_Corley @Kristine1G @amyklobuchar @Sty...       en
1                2  https://t.co/YJNiO0p1JV Flagstar Bank disclose...       en
2                4  OOPS. I typed her name incorrectly (today’s ...       en
3               14  @SaltBurned [When the first sign of surrender ...       en
4               27  "It's Your fight to save the UK from the liars...       en
...            ...                                                ...      ...
204144      937821  🔥 UNDER CONTRACT 🔥\n\nSo happy for my cl...       en
204145      937822  A little advice for us Browns fans on the Desh...       en
204146      937823  @nathaliejacoby1 My head is spinning. She is a...       en
204147      937840  @RepMTG @January6thCmte Make sure this is your...       en
204148      937848  John Wall’s contract is holy. That man was g...       en

[204149 rows x 3 columns]

###### Above dataset `data1`contains all the litigious tweets of having 204149 rows

In [5]:
negative=pd.read_csv(r"D:\Data Sets\ML Assignment NLP\dataset\negative.csv",encoding="latin-1")

In [6]:
negative.sample()

Unnamed: 0                                               Text Language
197362      705282  #Breaking Seven-time champion Serena Williams ...       en

In [7]:
negative.drop(["Unnamed: 0"],axis='columns',inplace=True)

In [8]:
negative

Text Language
0       #BadBunny: Como dos gotas de agua: Joven se di...       es
1       It sucks for me since I'm focused on the natur...       en
2       @en_font Treballar a l’obra a partir dels 19...       ca
3       Social Security. Constant political crises dis...       en
4       @FilmThePoliceLA A broken rib can puncture a l...       en
...                                                   ...      ...
262215  @SBeaberry @EmmaVigeland No I haven't because ...       en
262216  Y’all talking bad on that dad feeding his on...       en
262217  8.SSR &amp; Disha Salian case should be solved...       en
262218  *ACCIDENT:  Damage Only* - Raleigh Fire Depart...       en
262219                         I'm lost and I'm found but       en

[262220 rows x 2 columns]

###### Above dataset `data2`contains all the negative tweets of having 262220  rows

In [9]:
a = negative['Language'].unique()
print(a)

['es' 'en' 'ca' 'pt' 'fr' 'hi' 'nl' 'ja' 'in' 'tl' 'zh' 'ru' 'und' 'tr'
 'sv' 'de' 'ar' 'ta' 'ht' 'pl' 'et' 'da' 'ko' 'no' 'qme' 'th' 'uk' 'ro'
 'si' 'it' 'qam' 'cy' 'zxx' 'ml' 'ne' 'mr' 'cs' 'qst' 'eu' 'bn' 'gu' 'is'
 'qht' 'fa' 'fi' 'ckb' 'el' 'hu' 'ur' 'sl' 'lt' nan 'art' 'pa' 'vi' 'kn'
 'sr' 'lv' 'or' 'iw' 'te' 'bg' 'sd']


In [10]:
a = negative['Language'].unique()


In [11]:
positive=pd.read_csv(r"D:\Data Sets\ML Assignment NLP\dataset\negative.csv",encoding="latin-1")
positive.drop(["Unnamed: 0"],axis='columns',inplace=True)

In [12]:
positive

Text Language
0       #BadBunny: Como dos gotas de agua: Joven se di...       es
1       It sucks for me since I'm focused on the natur...       en
2       @en_font Treballar a l’obra a partir dels 19...       ca
3       Social Security. Constant political crises dis...       en
4       @FilmThePoliceLA A broken rib can puncture a l...       en
...                                                   ...      ...
262215  @SBeaberry @EmmaVigeland No I haven't because ...       en
262216  Y’all talking bad on that dad feeding his on...       en
262217  8.SSR &amp; Disha Salian case should be solved...       en
262218  *ACCIDENT:  Damage Only* - Raleigh Fire Depart...       en
262219                         I'm lost and I'm found but       en

[262220 rows x 2 columns]

###### Above dataset `data3`contains all the positive tweets of having positive rows

uncertainty=pd.read_csv(r"D:\Data Sets\ML Assignment NLP\dataset\uncertainty.csv",encoding="latin-1")

In [14]:
uncertainty.drop(['Unnamed: 0'],axis='columns',inplace=True)

In [15]:
uncertainty

a = uncertainty['Language'].unique()
print(a)


['en' 'fr' 'ca' 'es' 'in' 'ja' 'tr' 'zh' 'pt' 'sv' 'ar' 'it' 'tl' 'ko'
 'hu' 'ht' 'pl' 'nl' 'et' 'qme' 'zxx' 'und' 'de' 'th' 'hi' 'fi' 'ru' 'cy'
 'lv' 'da' 'qam' 'fa' 'no' 'sl' 'cs' 'ro' 'si' 'is' 'vi' 'ur' 'el' 'lt'
 'art' 'eu' 'qht' 'uk' 'ta' 'kn'
 "[Photo(previewUrl='https://pbs.twimg.com/media/EPYG2rKVAAA1e_O?format=jpg&name=small', fullUrl='https://pbs.twimg.com/media/EPYG2rKVAAA1e_O?format=jpg&name=large')]"
 'mr']


In [16]:
uncertainty

Text Language
0       @ShawnTarloff @itsmieu you can also relate thi...       en
1       @co_fuckoff J'espere que ça passera le plus v...       fr
2       ew its almost valentine’s day https://t.co/F...       en
3       @BlackMercury3 I’m thinking that maybe Ironw...       en
4       @zoidberg95 Infinite Diversity in Infinite Com...       en
...                                                   ...      ...
206935  There is a bag of cheese hidden somewhere in m...       en
206936  @Treg2Cole That’s a hellz nah for me. 🤢. ...       en
206937  @Sam_Mallory El Mizouni and Roles in the same ...       en
206938  If you don’t take risks you’ll always end ...       en
206939  @michaelharriot If Friends were better he migh...       en

[206940 rows x 2 columns]

###### Above dataset `data4`contains all the uncertainty tweets of having 206940 rows

### concatinating all the 4 dataframes

In [17]:
data=pd.concat([litigious,negative,positive,uncertainty])

In [18]:
data

Unnamed: 0                                               Text Language
0              0.0  @Charlie_Corley @Kristine1G @amyklobuchar @Sty...       en
1              2.0  https://t.co/YJNiO0p1JV Flagstar Bank disclose...       en
2              4.0  OOPS. I typed her name incorrectly (today’s ...       en
3             14.0  @SaltBurned [When the first sign of surrender ...       en
4             27.0  "It's Your fight to save the UK from the liars...       en
...            ...                                                ...      ...
206935         NaN  There is a bag of cheese hidden somewhere in m...       en
206936         NaN  @Treg2Cole That’s a hellz nah for me. 🤢. ...       en
206937         NaN  @Sam_Mallory El Mizouni and Roles in the same ...       en
206938         NaN  If you don’t take risks you’ll always end ...       en
206939         NaN  @michaelharriot If Friends were better he migh...       en

[935529 rows x 3 columns]

In [19]:
data=data.reset_index()

#### reseting all the index values as we have concatinated many dataframes to one 

In [20]:
data.drop(labels=['index','Unnamed: 0'],axis=1,inplace=True)

#### drop unnamed column

In [21]:
data['Language'].value_counts()

en                                                                                                                                                                      866940
fr                                                                                                                                                                       16023
es                                                                                                                                                                       11426
pt                                                                                                                                                                        9984
ja                                                                                                                                                                        5639
                                                                                                                             

In [54]:
import re
import emoji
from textblob import TextBlob
from nltk.tokenize import sent_tokenize,word_tokenize
from nltk.corpus import stopwords
from nltk.stem import PorterStemmer,LancasterStemmer,SnowballStemmer,WordNetLemmatizer

#### 1.Here we are using `regex` for removing unnecessary characters in text 
#### 2. `emoji` using this we are decoding the information from emoji present in the text
#### 3. `TextBlob` using textblob we are correcting the spelling of english words
#### 4. `Stem` using stemming we are storing the root word of a particular word 
#### 5. `tokenize` using this dividing paragraph or sentences  into tokens

In [22]:
data = data[data['Language']=='en']

#### convert to english

In [23]:
data

Text Language
0       @Charlie_Corley @Kristine1G @amyklobuchar @Sty...       en
1       https://t.co/YJNiO0p1JV Flagstar Bank disclose...       en
2       OOPS. I typed her name incorrectly (today’s ...       en
3       @SaltBurned [When the first sign of surrender ...       en
4       "It's Your fight to save the UK from the liars...       en
...                                                   ...      ...
935524  There is a bag of cheese hidden somewhere in m...       en
935525  @Treg2Cole That’s a hellz nah for me. 🤢. ...       en
935526  @Sam_Mallory El Mizouni and Roles in the same ...       en
935527  If you don’t take risks you’ll always end ...       en
935528  @michaelharriot If Friends were better he migh...       en

[866940 rows x 2 columns]

In [24]:
data.isnull().sum()

Text        0
Language    0
dtype: int64

In [25]:
stop_words = stopwords.words("english")
stop_words.remove('not')

#### removing `not` in the stopwords 

In [26]:
%%time
sentances = []
for i in data['Text']:
    word = []
    sent = i.lower()                           #first converting into lower case
    sent = emoji.demojize(sent)             #getting desired information from the emojis
    sent = re.sub('‘|’',"'",sent)         
    sent = re.sub('&amp;',"and",sent)
    sent = re.sub('<.*?>','',sent)                # removing unnecessary characters
    sent = re.sub('https?://\S*|www\S*','',sent)
    sent = re.sub(' @\w+| #\w+',"",sent)
    sent = re.sub("[^a-z0-9 ]","",sent)
    for i in word_tokenize(sent):                   #using word_tokenize dividing into words
        if i in stop_words:
            pass
        else:word.append(PorterStemmer().stem(i))  #neglecting the words which are in the stopwords
    sentances.append(" ".join(word))                 #Appending into the list

Wall time: 15min 23s


In [30]:
cleaned_text = pd.DataFrame({"cleaned_Text":sentances})

In [31]:
cleaned_text

cleaned_Text
0       charliecorley testimoni not evid court law sta...
1       flagstar bank disclos data breach impact 15mil...
2       oop type name incorrectli today brave wit 6 7 ...
3       saltburn first sign surrend come michael smile...
4       fight save uk liar cheat charlatan steve bray ...
...                                                   ...
866935           bag chees hidden somewher hous aint talk
866936                   treg2col that hellz nah mayb fri
866937  sammallori el mizouni role team could realli g...
866938  dont take risk youll alway end work someon kin...
866939  michaelharriot friend better might notic compa...

[866940 rows x 1 columns]

#### bag of words

In [ ]:
from sklearn.feature_extraction.text import CountVectorizer, TfidfVectorizer

#### using `CountVectorizer` converting text into vectors

In [32]:
count_obj=CountVectorizer()
trans=count_obj.fit_transform(cleaned_text['cleaned_Text'][0:25000])
trans

<25000x42934 sparse matrix of type '<class 'numpy.int64'>'
	with 351485 stored elements in Compressed Sparse Row format>

In [33]:
%%time
array_count = trans.toarray()

Wall time: 557 ms


In [34]:
count_obj.vocabulary_

{'charliecorley': 8158,
 'testimoni': 37685,
 'not': 27235,
 'evid': 13930,
 'court': 9969,
 'law': 22291,
 'state': 36153,
 'feder': 14557,
 'must': 26164,
 'stand': 36081,
 'cross': 10299,
 'examin': 13980,
 'flagstar': 14930,
 'bank': 5082,
 'disclos': 11953,
 'data': 10836,
 'breach': 6607,
 'impact': 18854,
 '15million': 439,
 'individu': 19085,
 'oop': 28075,
 'type': 39633,
 'name': 26325,
 'incorrectli': 19011,
 'today': 38663,
 'brave': 6584,
 'wit': 41862,
 'probabi': 30486,
 'dont': 12357,
 'forget': 15197,
 'request': 32143,
 'saltburn': 33380,
 'first': 14854,
 'sign': 34909,
 'surrend': 36917,
 'come': 9093,
 'michael': 24962,
 'smile': 35335,
 'split': 35883,
 'face': 14241,
 'like': 22938,
 'sharp': 34576,
 'knife': 21740,
 'flesh': 14964,
 'would': 42175,
 'chill': 8337,
 'dean': 11072,
 'core': 9778,
 'ice': 18590,
 'cold': 8999,
 'exist': 14059,
 'natur': 26446,
 'archangel': 4115,
 'enj': 13461,
 'fight': 14732,
 'save': 33615,
 'uk': 39728,
 'liar': 22823,
 'cheat'

In [35]:
pd.DataFrame(array_count,columns=sorted(count_obj.vocabulary_))

00  000  000001  0007  001  0047  004coinmarketcap  006  007  007anoth  \
0       0    0       0     0    0     0                 0    0    0         0   
1       0    0       0     0    0     0                 0    0    0         0   
2       0    0       0     0    0     0                 0    0    0         0   
3       0    0       0     0    0     0                 0    0    0         0   
4       0    0       0     0    0     0                 0    0    0         0   
...    ..  ...     ...   ...  ...   ...               ...  ...  ...       ...   
24995   0    0       0     0    0     0                 0    0    0         0   
24996   0    0       0     0    0     0                 0    0    0         0   
24997   0    0       0     0    0     0                 0    0    0         0   
24998   0    0       0     0    0     0                 0    0    0         0   
24999   0    0       0     0    0     0                 0    0    0         0   

       ...  zumwalt  zuneth  zuwairahnasir  zyagle09  zychaxrma  zygotes3  \
0      ...        0       0              0         0          0         0   
1      ...        0       0              0         0          0         0   
2      ...        0       0              0         0          0         0   
3      ...        0       0              0         0          0         0   
4      ...        0       0              0         0          0         0   
...    ...      ...     ...            ...       ...        ...       ...   
24995  ...        0       0              0         0          0         0   
24996  ...        0       0              0         0          0         0   
24997  ...        0       0              0         0          0         0   
24998  ...        0       0              0         0          0         0   
24999  ...        0       0              0         0          0         0   

       zylstra  zzztyler  zzzz  zzzzz  
0            0         0     0      0  
1            0         0     0      0  
2            0         0     0      0  
3            0         0     0      0  
4            0         0     0      0  
...        ...       ...   ...    ...  
24995        0         0     0      0  
24996        0         0     0      0  
24997        0         0     0      0  
24998        0         0     0      0  
24999        0         0     0      0  

[25000 rows x 42934 columns]

#### TF-IDF

In [37]:
tf= TfidfVectorizer()
trans_tf = tf.fit_transform(cleaned_text['cleaned_Text'][0:5000])
trans_tf

<5000x14164 sparse matrix of type '<class 'numpy.float64'>'
	with 69322 stored elements in Compressed Sparse Row format>

In [41]:
pd.DataFrame(trans_tf.toarray(),columns=sorted(tf.vocabulary_.keys()))

0047  007   01  02906for  030  033  036  038crypto  042   05  ...  \
0      0.0  0.0  0.0       0.0  0.0  0.0  0.0        0.0  0.0  0.0  ...   
1      0.0  0.0  0.0       0.0  0.0  0.0  0.0        0.0  0.0  0.0  ...   
2      0.0  0.0  0.0       0.0  0.0  0.0  0.0        0.0  0.0  0.0  ...   
3      0.0  0.0  0.0       0.0  0.0  0.0  0.0        0.0  0.0  0.0  ...   
4      0.0  0.0  0.0       0.0  0.0  0.0  0.0        0.0  0.0  0.0  ...   
...    ...  ...  ...       ...  ...  ...  ...        ...  ...  ...  ...   
4995   0.0  0.0  0.0       0.0  0.0  0.0  0.0        0.0  0.0  0.0  ...   
4996   0.0  0.0  0.0       0.0  0.0  0.0  0.0        0.0  0.0  0.0  ...   
4997   0.0  0.0  0.0       0.0  0.0  0.0  0.0        0.0  0.0  0.0  ...   
4998   0.0  0.0  0.0       0.0  0.0  0.0  0.0        0.0  0.0  0.0  ...   
4999   0.0  0.0  0.0       0.0  0.0  0.0  0.0        0.0  0.0  0.0  ...   

      zoomcarindia  ztpetrizzo   zu  zubac  zubair  zuneth  zuwairahnasir  \
0              0.0         0.0  0.0    0.0     0.0     0.0            0.0   
1              0.0         0.0  0.0    0.0     0.0     0.0            0.0   
2              0.0         0.0  0.0    0.0     0.0     0.0            0.0   
3              0.0         0.0  0.0    0.0     0.0     0.0            0.0   
4              0.0         0.0  0.0    0.0     0.0     0.0            0.0   
...            ...         ...  ...    ...     ...     ...            ...   
4995           0.0         0.0  0.0    0.0     0.0     0.0            0.0   
4996           0.0         0.0  0.0    0.0     0.0     0.0            0.0   
4997           0.0         0.0  0.0    0.0     0.0     0.0            0.0   
4998           0.0         0.0  0.0    0.0     0.0     0.0            0.0   
4999           0.0         0.0  0.0    0.0     0.0     0.0            0.0   

      zychaxrma  zygotes3  zylstra  
0           0.0       0.0      0.0  
1           0.0       0.0      0.0  
2           0.0       0.0      0.0  
3           0.0       0.0      0.0  
4           0.0       0.0      0.0  
...         ...       ...      ...  
4995        0.0       0.0      0.0  
4996        0.0       0.0      0.0  
4997        0.0       0.0      0.0  
4998        0.0       0.0      0.0  
4999        0.0       0.0      0.0  

[5000 rows x 14164 columns]

in countvectorizer (bag of words) and tfidfvectorizer(TF-IDF) i have reduced datapoints because of memory problem so i have limited to 25000 rows in countvectorizer instead of 866940 and 5000 in tfidfvectorizer